### PR TITLE
Rules flyout second check in

### PR DESCRIPTION
**Pull request by isaac73**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

Filled out all dropdown in rules flout
...

**Checklist:**

- [x] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: public/locales/en/translations.json [hunk 1]
```diff
@@ -57,15 +57,19 @@
       "field": "Field",
       "fieldPlaceholder": "Select field",
       "calculation": "Calculation",
-      "calculationPlaceholder": "select",
+      "calculationPlaceholder": "Select",
```
Comment (mechaffin) [inline public/locales/en/translations.json]: Move the rulesFlyout translations to be under the page they go with.  So in this case.... under the rules, then under .flyouts.
See devices.flyouts.[flyoutname] for examples.

File: src/app.config.js [hunk 2]
```diff
@@ -30,7 +30,15 @@ const Config = {
     return validExtensions.indexOf('.' + fileExt) > -1;
   },
   emptyValue: '--',
-  maxTopAlarms: 5
+  maxTopAlarms: 5,
+
+  OPERATOR_OPTIONS: [
```
Comment (mechaffin) [inline src/app.config.js]: Check with Molly.  Both of you have something similar to this.  If they are the same, then we should utilize the same code... and certainly not completely different approaches.

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js [hunk 3]
```diff
@@ -131,48 +211,58 @@ export class RuleNew extends LinkedComponent {
...
                       <FormLabel isRequired="true">{t(`rulesFlyout.condition.value`)}</FormLabel>
                       <FormControl
                         type="text"
                         placeholder={t(`rulesFlyout.condition.valuePlaceholder`)}
-                        link={condition.value} />
+                        link={condition.valueLink} />
                     </FormGroup>
+                    {
+                      conditionLinks.length > 1 &&
+                      <Btn className="delete-btn" svg={svgs.trash} onClick={this.deleteCondition(idx)}>Delete</Btn>
```
Comment (mechaffin) [inline src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: translate

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js [hunk 4]
```diff
@@ -131,48 +211,58 @@ export class RuleNew extends LinkedComponent {
...
                     <FormGroup>
                       <FormLabel isRequired="true">{t(`rulesFlyout.condition.value`)}</FormLabel>
                       <FormControl
                         type="text"
                         placeholder={t(`rulesFlyout.condition.valuePlaceholder`)}
-                        link={condition.value} />
+                        link={condition.valueLink} />
                     </FormGroup>
+                    {
+                      conditionLinks.length > 1 &&
```
Comment (mechaffin) [inline src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: The button should probably enable/disable instead of showing/hiding.

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js [hunk 5]
```diff
@@ -184,23 +274,23 @@ export class RuleNew extends LinkedComponent {
                   <Radio
                     placeholder="critical"
                     checked={this.state.severityLevel.critical}>
-                    <SeverityRenderer value="Critical" context={{ t }} iconOnly={false} />
+                    <SeverityRenderer value={'Critical'} context={{ t }} iconOnly={false} />
```
Comment (mechaffin) [inline src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: Are these magical strings? Perhaps add constants somewhere?

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.scss [hunk 6]
```diff
@@ -39,4 +39,10 @@
     @include rem-fallback(left, 180px);
     @include rem-fallback(top, 42px);
   }
+
+  .delete-btn {
+    @include rem-fallback(margin-left, 0px);
```
Comment (mechaffin) [inline src/components/pages/rules/flyouts/ruleNew/ruleNew.scss]: It shouldn't be necessary for 0px to use rem-fallback.  
This should be sufficient:
margin-left: 0;

File: src/components/pages/rules/rules.js [hunk 7]
```diff
@@ -56,7 +56,7 @@ export class Rules extends Component {
   getSoftSelectId = ({ id }) => id;
 
   render () {
-    const { t, rules, error, isPending, lastUpdated, entities, fetchRules } = this.props;
+    const { t, rules, error, isPending, lastUpdated, entities, fetchRules, deviceGroups } = this.props;
```
Comment (mechaffin) [inline src/components/pages/rules/rules.js]: too many on a single line.  separate into one per line.

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 4 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: I think otherwise. There is no use of displaying it when there is only one condition.

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 5 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: no need for constants since we are not reusing them anywhere else

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js [hunk 8]
```diff
@@ -60,158 +62,248 @@ export class RuleNew extends LinkedComponent {
...
+      const valueLink = conditionLink.forkTo('value');
+      const durationLink = conditionLink.forkTo('duration');
+      return { fieldLink, calculationLink, operatorLink, valueLink, durationLink };
     });
 
     return (
       <Flyout.Container>
         <Flyout.Header>
-          <Flyout.Title>{t(`rulesFlyout.newRule`)}</Flyout.Title>
+          <Flyout.Title>{t(`${tPath}newRule`)}</Flyout.Title>
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: I don't think it will be difficult to search. Ex: if you are searching "newRule" string you'll still find it. Otherwise we have to put "rules.flyouts.ruleEditor.xxx" every where.

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js [hunk 9]
```diff
@@ -60,158 +62,248 @@ export class RuleNew extends LinkedComponent {
...
   }
 
+  onGroupIdChange = (selectedItem) => {
+    this.getDeviceCountAndFields(selectedItem.target.value.value);
+  }
+
+  getDeviceCountAndFields(groupId) {
+    this.props.deviceGroups.forEach(group => {
+      if (group.id === groupId) {
+        this.subscription = IoTHubManagerService.getDevices(group.Conditions)
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: You won't get the updated values if you added new device properties in a different browser tab unless you close and reopen flyout. I think availability is more important than performance here. Also I can't think why a user would frequently change dropdown values.

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+} from 'utilities';
+import Flyout from 'components/shared/flyout';
+import Config from 'app.config';
+import './ruleEditor.css';
+import { IoTHubManagerService } from 'services';
+
+const Section = Flyout.Section;
+const ruleNameValidator = (new Validator()).check(Validator.notEmpty, 'Name is required');
+const deviceGroupValidator = (new Validator()).check(Validator.notEmpty, 'Device group is required');
+const tPath = "rules.flyouts.ruleEditor.";
```
Comment (ppathan) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: nit: single quotes

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+const severityLevels = ["critical", "warning", "info"];
```
Comment (ppathan) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: nit: single quotes

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+const calculations = ["average", "instant"];
```
Comment (ppathan) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: nit: single quotes

File: public/locales/en/translations.json (hunk 1 above)
Comment (isaac73) [reply public/locales/en/translations.json]: fixed

File: src/app.config.js (hunk 2 above)
Comment (isaac73) [reply src/app.config.js]: fixed

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 3 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: fixed

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 4 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: fixed

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 5 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: fixed

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.scss (hunk 6 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.scss]: fixed

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js (hunk 10 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: fixed all single quotes

File: src/components/pages/rules/flyouts/ruleNew/ruleNew.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleNew/ruleNew.js]: fixed all comments

File: src/services/models/telemetryModels.js [hunk 13]
```diff
@@ -63,3 +63,33 @@ export const toMessagesModel = (response = {}) => (response.items || [])
...
+  export const toNewRuleRequestModel = ({
+    name,
+    description,
+    groupId,
+    conditions,
+    severity,
+    enabled
+  }) => {
+    const Conditions = [];
+    conditions.map((condition) =>{
```
Comment (spryor) [inline src/services/models/telemetryModels.js]:
```
const Conditions = conditions.map(condition => ({
  Field: condition.field,
  Operator: condition.operator,
  ...
}));
```

File: src/services/models/telemetryModels.js [hunk 14]
```diff
@@ -63,3 +63,33 @@ export const toMessagesModel = (response = {}) => (response.items || [])
...
+      let modelCondition = {
+        Field: condition.field,
+        Operator: condition.operator,
+        Calculation: condition.calculation,
+        Duration: condition.duration,
+        Value: condition.value,
```
Comment (spryor) [inline src/services/models/telemetryModels.js]: Remove the `,` at the end here

File: src/services/models/telemetryModels.js [hunk 15]
```diff
@@ -63,3 +63,33 @@ export const toMessagesModel = (response = {}) => (response.items || [])
...
+      Conditions.push(modelCondition);
+      return condition;
+    });
+    return {
+      Name: name,
+      Description: description,
+      GroupId: groupId,
+      Severity: severity,
+      Enabled: enabled,
+      Conditions: Conditions
```
Comment (spryor) [inline src/services/models/telemetryModels.js]:
```
Enabled: enabled,
Conditions
```

File: src/services/telemetryService.js [hunk 16]
```diff
@@ -22,6 +22,18 @@ export class TelemetryService {
       .map(toRulesModel);
   }
 
+  /** creates a new rule */
+  static createRule(rule = {}) {
```
Comment (spryor) [inline src/services/telemetryService.js]: Remove the default values of `{}`

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+              this.setState({
+                fieldOptions: this.getConditionFields(groupDevices),
+                devicesAffected: groupDevices.length
+              });
+            },
+            error => this.setState({ error })
+          );
+        return true;
+      }
+      return false;
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: eslint error if removed

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+                </FormGroup>
+              </Section.Content>
+              <Section.Content>
+                <FormGroup>
+                  <FormLabel>{t('rules.flyouts.ruleEditor.ruleStatus')}</FormLabel>
+                  <ToggleBtn value={this.state.formData.enabled}>{this.state.formData.enabled ? t('rules.flyouts.ruleEditor.ruleEnabled') : t('rules.flyouts.ruleEditor.ruleDisabled')}</ToggleBtn>
+                </FormGroup>
+              </Section.Content>
+            </Section.Container>
+            <Section.Container collapsable={false}>
```
[the remark this answers is not in the document]
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: `SummarySection` seems like a overkill. Also style is a bit different

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  onGroupIdChange = ({ target: { value: { value = {} } } }) => {
+    this.getDeviceCountAndFields(value);
+  }
+
+  getDeviceCountAndFields(groupId) {
+    this.props.deviceGroups.some(group => {
+      if (group.id === groupId) {
+        this.subscription = IoTHubManagerService.getDevices(group.conditions)
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: Need to check if the subscription has anything in it before overriding it. If something is there it should be unsubscribed.

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  render() {
+    const { onClose, t } = this.props;
+    const calculationOptions = calculations.map(value => ({
+      label: t(`rules.flyouts.ruleEditor.calculation.${value}`),
+      value
+    }));
+    // Validators
+    const requiredValidator = (new Validator()).check(Validator.notEmpty, () => this.props.t('rules.flyouts.ruleEditor.validation.required'));
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: Don't need the function as the message here since you area already in the render method.

If you keep the function and move this to the constructor that should work too.

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+    this.deviceGroupLink = this.formDataLink.forkTo('groupId')
+      .map(({ value }) => value)
+      .withValidator(requiredValidator);
+    this.conditionsLink = this.formDataLink.forkTo('conditions');
+    this.severityLink = this.formDataLink.forkTo('severity');
+    //todo toggle button didn't support link
+    this.enabledLink = this.formDataLink.forkTo('enabled');
+    // Create the state link for the dynamic form elements
+    const conditionLinks = this.conditionsLink.getLinkedChildren(conditionLink => {
+      const fieldLink = conditionLink.forkTo('field').map(({ value }) => value).withValidator(requiredValidator);;
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: Remove the `;;`

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+                    <FormGroup>
+                      <FormLabel isRequired='true'>{t('rules.flyouts.ruleEditor.condition.value')}</FormLabel>
+                      <FormControl
+                        type='text'
+                        placeholder={t('rules.flyouts.ruleEditor.condition.valuePlaceholder')}
+                        link={condition.valueLink} />
+                    </FormGroup>
+                    {
+                      conditionLinks.length > 1 &&
+                      <Btn className='padded-top flyout-btns' svg={svgs.trash} onClick={this.deleteCondition(idx)}>{t('rules.flyouts.ruleEditor.delete')}</Btn>
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: `flyout-btns` needs to be removed

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+  .rule-property-container {
+    border: none;
+    @include rem-fallback(margin-top, 20px);
+  }
+
+  .devices-affected{
+    display: flex;
+    flex-flow: row nowrap;
+    position: relative;
+    @include rem-fallback(top, 18.5px);
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss]: Can't have `18.5px`

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+@import 'src/styles/mixins';
+
+.new-rule-flyout-container {
+
+  .rule-property-container {
+    border: none;
+    @include rem-fallback(margin-top, 20px);
+  }
+
+  .devices-affected{
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss]: `ed[space]{`

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+    display: flex;
+    flex-flow: row nowrap;
+    position: relative;
+    @include rem-fallback(top, 18.5px);
+    @include rem-fallback(height, 67px);
+  }
+
+  @include themify($themes) {
+    .devices-affected-dynamic {
+      @include rem-fallback(font-size, 48px);
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss]: Place the @includes below explicit values.
Also, always define the theme mixin at the bottom of the file.

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+      font-weight: bold;
+      color: themed('colorHighest');
```
Comment (spryor) [inline src/components/pages/rules/flyouts/ruleEditor/ruleEditor.scss]: `.devices-affected-dynamic` should be defined above with static props while the themable props are only in the themify.

File: src/services/models/configModels.js [hunk 27]
```diff
@@ -28,7 +28,7 @@ export const prepareLogoResponse = (responseWrapper) => {
 }
 
 export const toSolutionSettingThemeModel = (response = {}) => reshape(response, {
-  'azureMapsKey': 'azureMapsKey',
+  'AzureMapsKey': 'azureMapsKey',
```
Comment (spryor) [inline src/services/models/configModels.js]: Undo this change

File: src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js (hunk 19 above)
Comment (isaac73) [reply src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js]: FYI - not done this way in new device flyout. I'll do here.
